### ISSUE TITLE
Consider skipping `bazel_tools@_` from lockfile

### ISSUE THREAD

**Issue by layus** [bazelbuild/bazel#19971]
### Description of the bug:

MODULE.bazel.lock files contain entries for `bazel_tools` builtin module, as well as all its dependencies. Because that module is builtin, its version changes with bazel releases, meaning that the lockfile gets invalidated when built with a different bazel version. Meaning that the bazel version is an integral part of the build dependencies.

It makes sense to keep bazel_tools and its dependencies in the lockfile if they take part of the modules version resolution. But maybe it would be better to keep them out of the modules version selection.

Overall, the question is maybe about compatibility between bazel versions. Are projects supposed to build only with a single bazel version as .bazelversion suggests ? In my context, being able to build a bazel project with different bazel versions makes things simpler.

Is there an official policy about bazel versions compatibility ?

### Which category does this issue belong to?

Core, Documentation

### Which operating system are you running Bazel on?

NixOS

### What is the output of `bazel info release`?

7.0.0rc2

### If `bazel info release` returns `development version` or `(@non-git)`, tell us how you built Bazel.

https://github.com/NixOS/nixpkgs/pull/262152

**Comment by fmeum** [comment]
@Pavank1992 Could you reassign to team External-Deps?

**Comment by meteorcloudy** [comment]
> Meaning that the bazel version is an integral part of the build dependencies.

This is true and exactly the reason why those builtin modules have to be recorded in the lockfile to ensure correctness.

But I don't think this will break compatibilities for building with different Bazel versions, it just means the lockfile could be updated if you change the bazel version (unless if you enable `--lockfile_mode=error`).

As the starlarkification of native build rules progresses, we should be able to reduce the dependencies required by the `@bazel_tools` builtin modules. And hopefully one day, we won't need the builtin module at all.

For keeping Bazel compatibilities:

- If you are a rule author, there is some general recommendations : https://bazel.build/release/rule-compatibility#best-practices
- If it's for an end project, it really depends on your own use cases.

**Comment by meteorcloudy** [comment]
Do you actually need to switch Bazel versions frequently? If so, we would like to understand why.

**Comment by meteorcloudy** [comment]
Oh, that is interesting. We kind of have to solve similar problems for running integration tests on Bazel CI. We want to make sure we can run Bazel integration tests without internet access, therefore we need to provide a pre-populated repository cache for the inner Bazel inside the test by building the [`//src:test_repos`](https://github.com/bazelbuild/bazel/blob/f898da56995809140b65d2b3b6f5b631a3ec511a/src/BUILD#L562C13-L562C23) in advance in the build step. 

You might also be interested in https://github.com/bazelbuild/bazel/pull/20042

Some links:
- https://github.com/bazelbuild/bazel/blob/f898da56995809140b65d2b3b6f5b631a3ec511a/.bazelci/presubmit.yml#L20-L20
- https://github.com/bazelbuild/bazel/blob/f898da56995809140b65d2b3b6f5b631a3ec511a/.bazelrc#L71-L73
- https://github.com/bazelbuild/bazel/blob/2052c635ecc41d9bf1b8236ebfed6399b2c28a57/src/test/shell/testenv.sh.tmpl#L327-L330
